include/api/CGlobalIdDataCategorizer.h

```diff
@@ -0,0 +1,129 @@
+//! categorizer.  An extra level cannot be introduced now, as that
+//! would break backwards compatibility of the state.
+//!
+class API_EXPORT CGlobalIdDataCategorizer {
+public:
+    //! Function used for persisting objects of this class
+    using TPersistFunc = std::function<void(core::CStatePersistInserter&)>;
+
+public:
+    CGlobalIdDataCategorizer(std::string partitionFieldName,
+                             model::CDataCategorizer::TDataCategorizerPtr dataCategorizer,
+                             CCategoryIdMapper::TCategoryIdMapperPtr categoryIdMapper);
+
+    //! Dump stats
+    void dumpStats() const;
+
+    //! Compute a category from a string.  The raw string length may be longer
+    //! than the length of the passed string, because the passed string may
+    //! have the date stripped out of it.  If the category changes as a result
+    //! write it to the output.
+    CGlobalCategoryId
+    computeAndUpdateCategory(bool isDryRun,
+                             const model::CDataCategorizer::TStrStrUMap& fields,
+                             const std::string& messageToCategorize,
+                             const std::string& rawMessage,
+                             model::CResourceMonitor& resourceMonitor,
+                             CJsonOutputWriter& jsonOutputWriter);
+
+    //! Make a function that can be called later to persist state in the
+    //! foreground, i.e. in the knowledge that no other thread will be
+    //! accessing the data structures this method accesses.
+    TPersistFunc makeForegroundPersistFunc() const;
+
+    //! Make a function that can be called later to persist state in the
+    //! background, i.e. copying any required data such that other threads
+    //! may modify the original data structures while persistence is taking
+    //! place.
+    TPersistFunc makeBackgroundPersistFunc() const;
+
+    //! Populate the object from part of a state document
+    bool acceptRestoreTraverser(core::CStateRestoreTraverser& traverser);
+
+    //! Access to the categorizer key.
+    const std::string& categorizerKey() const {
+        return m_CategoryIdMapper->categorizerKey();
+    }
+
+    //! Writes out to the JSON output writer any category that has changed
+    //! since the last time this method was called.
+    void writeOutChangedCategories(CJsonOutputWriter& jsonOutputWriter);
+
+    //! Force an update of the resource monitor.
+    void forceResourceRefresh(model::CResourceMonitor& resourceMonitor);
+
+private:
+    //! Used by deferred persistence functions
+    static void acceptPersistInserter(const model::CDataCategorizer::TPersistFunc& dataCategorizerPersistFunc,
+                                      const model::CCategoryExamplesCollector& examplesCollector,
+                                      const CCategoryIdMapper& categoryIdMapper,
+                                      core::CStatePersistInserter& inserter);
+
+private:
```

Comment:
Is there any reason not to cache the GlobalId on this class. This seems like it would do away with all the map lookups.

Reply:
Yes, since the changes of 69cc35c9a62a8e76922acbf2212f9b822038de25 `CPerPartitionCategoryIdMapper` just does local -> global mapping for a single partition, so there are no map lookups within it.